sympy/algebras/quaternion.py

```diff
@@ -78,8 +94,17 @@ class Quaternion(Expr):
 
     is_commutative = False
 
-    def __new__(cls, a=0, b=0, c=0, d=0, real_field=True):
-        a, b, c, d = map(sympify, (a, b, c, d))
+    def __new__(cls, a=0, b=0, c=0, d=0, real_field=True, norm=None):
+        a, b, c, d, norm = map(sympify, (a, b, c, d, norm))
+
+        if (norm is not None) and (norm.is_Number):
+            # if all elements are numerical
+            if (all(i.is_Number is True for i in [a, b, c, d]) and
+                    norm**2 != a**2 + b**2 + c**2 + d**2):
```

Comment:
```suggestion
        if norm is not None and norm**2 != a**2 + b**2 + c**2 + d**2:
```

Reply:
So how about:

```python
if norm is not None and norm.is_number and all(i.is_number for i in (a,b,c,d)):
    if norm**2 != a**2 + b**2 + c**2 + d**2:
        raise ValueError("incompatible value for norm")
    if norm.is_zero:
        ...
```

Reply:
The problem with this, on the other hand, is that it will only test if `norm = 0` if `a,b,c,d` are also numbers

Reply:
```python
if norm is not None:
    if norm.is_zero:
        ...
    dif = norm**2 - (a**2 + b**2 + c**2 + d**2)
    if dif.is_number and not dif.is_zero:
        raise ValueError("incompatible value for norm")
```

Reply:
Instead of `norm.is_zero` I think I should stick to `not norm.is_positive` in order to safe check against negative numbers.

On the other hand, in your version we will be computing `dif` before checking if everything is a number, but if something isn't a number, then this calculation is useless... 

What do you think of this, to make the code a bit cleaner:
```
def set_norm(self, norm):
    norm = sympify(norm)

    if norm is not None and norm.is_number:
        if not norm.is_positive:
            raise ValueError("Input norm must be positive.")

        numerical = all(elem.is_number is True for elem in self.args)
        if (numerical and norm**2 != self.norm()**2):
            raise ValueError("Incompatible value for norm.")

    self._norm = norm
```

Reply:
That looks good. So is the call to `self.norm()` going to set `self._norm` which then gets overwritten by the `norm` that was passed? (Just checking my undersanding.)

Reply:
Exactly! And it is called in `__new__` too:
```
            obj._c = c
            obj._d = d
            obj._real_field = real_field
            obj._norm = None
            obj.set_norm(norm)
            return obj
```